File: q0230/q0230.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    int kthSmallest(TreeNode *root, int k)
    {
        stack<TreeNode *> node_stack;
        while (true)
        {
            // go all the way to the left most element
            while (root)
            {
                node_stack.push(root);
                root = root->left;
            }
            root = node_stack.top();
            k--;
            if (k == 0)
                return root->val;
            node_stack.pop();
            root = root->right;
        }
    }
};

File: q0230/q0230.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) {
        LinkedList<TreeNode> node_stack = new LinkedList<>();
        while (true) {
            while (root != null) {
                node_stack.push(root);
                root = root.left;
            }
            root = node_stack.pop();
            k--;
            if (k == 0)
                return root.val;
            root = root.right;
        }
    }
}